### ISSUE TITLE
A simple command to wrap any folder as a nix channel

### ISSUE THREAD

**Issue by CMCDragonkai** [NixOS/nixpkgs#14698]
Currently nix-channel expects a URI to a resource containing 2 things. A `nixexprs.tar.xz` and a file `binary-cache-url`.

There doesn't seem to be a `nix-*` command that can easily and automatically take any nix expression folder or expression and wrapping it up, satisfying the channel format.

I reckon such a command would be quite useful. And especially since nix-channel can use `file://` protocol for local filesystem. This would make custom channels super easy to create.

I would also prefer nix-channel to verify the signature of the channel assuming I have it available with already.

I'll be looking to create this command just using bash script. But this might be useful to integrate into the existing nix utilities. In which case will require C++ instead.

**Comment by joachifm** [comment]
I wonder, what is the reason for wanting this instead of setting `NIX_PATH=nixpkgs=`? Also, I don't think this is a nixpkgs issue per se.

**Comment by CMCDragonkai** [comment]
You can then allow other people to use it (over the net), and setup a generation timeline for a local package set.

**Comment by CMCDragonkai** [comment]
Oh and nix-env doesn't respect NIX_PATH.

**Comment by joachifm** [comment]
In what sense does it not?

**Comment by joachifm** [comment]
Are you aware that `nix-env -f` accepts any URL pointing to a nixexprs archive? If you need to use different or additional binary caches, that is covered by nix.conf or options (depending on your setup). To me, channels appear to be all but obsolete. Then again, I have simple needs (I asked why you needed this out of curiousity, not to dismiss the issue, just to be clear).

**Comment by CMCDragonkai** [comment]
Here
http://lethalman.blogspot.com.au/2014/09/nix-pill-15-nix-search-paths.html?m=1

**Comment by CMCDragonkai** [comment]
Yea so I know about -f. Channels appear to be a moving target kind of
package set. Similar to rolling releases.

**Comment by ericsagnes** [comment]
The proposal for the [new command line tool](https://gist.github.com/edolstra/efcadfd240c2d8906348) might interest you. 
There is a mention about dropping the channels and use `$NIX_PATH` instead.

**Comment by CMCDragonkai** [comment]
Yep, I've commented on that already. However using `NIX_PATH` is quite a different thing from channels. Channels represent a moving target of sorts, similar to other distros, but enhanced with generation timelines. Channels also allow merging of different channels, which is again similar to things like having a PPA in Ubuntu. It's a different kind of workflow. Not content addressable, but still useful for not-so-technical users who are coming from other distros.

**Comment by layus** [comment]
@CMCDragonkai I generate a channel-like tarball with 

```
$ nix-build nixos/release.nix -A channel --arg nixpkgs  '{ outPath = ./. ; revCount = "'$(git rev-list HEAD | wc -l)'"; shortRev = "'$(git rev-parse --short HEAD)'"; }'
...
/nix/store/hash-name/
$ sudo nix-channel --remove nixos
$ sudo nix-channel --add file:///nix/store/hash-name/tarballs/thetarball.tar.xz nixos
$ sudo nix-channel --update
```

I am quite sure this is not perfect (the tarball is unpacked then re-packed), but it gives you a hint of how it could work.

Otherwise, using hydra seems the canonical way to go.

**Comment by ZerataX** [comment]
still important to me

**Comment by yanganto** [comment]
@ZerataX I am a new user for NixOS, and also encounter the similar problem.  Here is the script I use, it maybe good for you too.
https://github.com/yanganto/parachutes/blob/master/bin/nix-folder2channel
